### PR TITLE
Sprint4 task2

### PR DESCRIPTION
**Pull request by iAcay**
The code which is below was implemented in order to meet the acceptance criterias of second task in sprint no. 4.

AC in this task:
-- An employee can buy a reward using their points
-- An employee can't buy a reward they can't afford
-- Buying a reward reduces the employee's points by the reward's price
-- There's an `Order` model
-- There's a system spec for showing a lower number of points after buying a reward

In this task, I decided to:

1) introduce many-to-many association between employees and rewards through orders. This solution guarantee possibility for employees to have many rewards and vice versa.

2)  implement the buy_reward method which handles buying rewards, decreasing earned points after bought and reject attempt to buy if employee's points are not enough.

3) make tests for decreasing number of points after buying a reward and for rejecting attempt to buy if the amount of points is not enough.

4) move tests for increasing and decreasing points after working on kudos into kudos_spec.rb file based on Mateusz suggestion in last CR.

### CODE REVIEW

File: app/controllers/rewards_controller.rb [hunk 1]
```diff
@@ -9,6 +9,18 @@ def show
     render :show, locals: { reward: reward }
   end
 
+  def buy_reward
```
Comment (mdobosz3) [inline app/controllers/rewards_controller.rb]: IMO this method should go to the order controller, finally you create order here. 
A good solution is also to use transactions when you perform several operations in one method.

File: config/routes.rb [hunk 2]
```diff
@@ -11,7 +11,11 @@
   devise_for :employees, path: 'employees', controllers: { sessions: "employees/sessions", registrations: "employees/registrations" }
 
   resources :kudos
-  resources :rewards, only: [:index, :show]
+  resources :rewards, only: [:index, :show] do
+    collection do
+      post 'buy_reward/:id', to: 'rewards#buy_reward', as: 'buy_reward'
```
Comment (mdobosz3) [inline config/routes.rb]: I would simplify the path to buying a reward:
```suggestion
      resources :orders, only: %i[index create]
```

**Review by mdobosz3 — COMMENTED** [review]
Hi, congratulations on making another task :), below are my suggestions for thought.

File: app/controllers/rewards_controller.rb (hunk 1 above)
Comment (iAcay) [reply app/controllers/rewards_controller.rb]: Understood :) Therefore I created orders_controller and next I moved method which handles buying rewards into this controller. Thank You :)

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def create
+    @reward = Reward.find(params[:order][:reward_id])
+    if current_employee.earned_points < @reward.price
+      redirect_to rewards_path, notice: "This reward is too expensive for you.
+                                          You need #{@reward.price - current_employee.earned_points} points more."
+    else
+      @order = Order.new(order_params)
+      if @order.save && current_employee.decrement(:earned_points, @reward.price).save
```
Comment (oskarlakner) [inline app/controllers/orders_controller.rb]: This introduces a problem - what if `current_employee.decrement(:earned_points, @reward.price).save` fails? You end up with an already saved order. Think about using a transaction

File: db/migrate/20220608184748_create_orders.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class CreateOrders < ActiveRecord::Migration[6.1]
+  def change
+    create_table :orders do |t|
+      t.bigint :employee_id
```
Comment (oskarlakner) [inline db/migrate/20220608184748_create_orders.rb]: you miss fk and indexes here
Screenshot 2022-06-12 at 08 51 37

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (iAcay) [reply app/controllers/orders_controller.rb]: Okay, I get that. Thanks :) I have already implemented transaction into this method.

File: db/migrate/20220608184748_create_orders.rb (hunk 4 above)
Comment (iAcay) [reply db/migrate/20220608184748_create_orders.rb]: Thanks. Corrected ;)

**Pull request merged**